packages/tldraw/src/lib/ui/hooks/useTldrawUiComponents.tsx

```diff
@@ -0,0 +1,104 @@
+import { ComponentType, createContext, useContext, useMemo } from 'react'
+import { DefaultKeyboardShortcutsDialog } from '../components/KeyboardShortcutsDialog/DefaultKeyboardShortcutsDialog'
+import { DefaultKeyboardShortcutsDialogContent } from '../components/KeyboardShortcutsDialog/DefaultKeyboardShortcutsDialogContent'
+import { DefaultMinimap } from '../components/Minimap/DefaultMinimap'
+import { DefaultNavigationPanel } from '../components/NavigationPanel/DefaultNavigationPanel'
+import { DefaultPageMenu } from '../components/PageMenu/DefaultPageMenu'
+import {
+	DefaultStylePanel,
+	type TLUiStylePanelProps,
+} from '../components/StylePanel/DefaultStylePanel'
+import {
+	DefaultStylePanelContent,
+	TLUiStylePanelContentProps,
+} from '../components/StylePanel/DefaultStylePanelContent'
+import { DefaultToolbar } from '../components/Toolbar/DefaultToolbar'
+import { DefaultActionsMenu } from '../components/menus/ActionsMenu/DefaultActionsMenu'
+import { DefaultActionsMenuContent } from '../components/menus/ActionsMenu/DefaultActionsMenuContent'
+import {
+	DefaultContextMenu,
+	TLUiContextMenuProps,
+} from '../components/menus/ContextMenu/DefaultContextMenu'
+import { DefaultContextMenuContent } from '../components/menus/ContextMenu/DefaultContextMenuContent'
+import { DefaultHelpMenu } from '../components/menus/HelpMenu/DefaultHelpMenu'
+import { DefaultHelpMenuContent } from '../components/menus/HelpMenu/DefaultHelpMenuContent'
+import { DefaultMainMenu } from '../components/menus/MainMenu/DefaultMainMenu'
+import { DefaultMainMenuContent } from '../components/menus/MainMenu/DefaultMainMenuContent'
+import { DefaultZoomMenu } from '../components/menus/ZoomMenu/DefaultZoomMenu'
+import { DefaultZoomMenuContent } from '../components/menus/ZoomMenu/DefaultZoomMenuContent'
+import { TLUiDialogProps } from './useDialogsProvider'
+
+export interface BaseTLUiComponents {
+	ContextMenu: ComponentType<TLUiContextMenuProps>
+	ContextMenuContent: ComponentType
+	ActionsMenu: ComponentType
+	ActionsMenuContent: ComponentType
+	HelpMenu: ComponentType
+	HelpMenuContent: ComponentType
+	ZoomMenu: ComponentType
+	ZoomMenuContent: ComponentType
+	MainMenu: ComponentType
+	MainMenuContent: ComponentType
+	Minimap: ComponentType
+	StylePanel: ComponentType<TLUiStylePanelProps>
+	StylePanelContent: ComponentType<TLUiStylePanelContentProps>
+	PageMenu: ComponentType
+	NavigationPanel: ComponentType
+	Toolbar: ComponentType
+	KeyboardShortcutsDialog: ComponentType<TLUiDialogProps>
+	KeyboardShortcutsDialogContent: ComponentType
+}
+
+/** @public */
+export type TLUiComponents = Partial<{
+	[K in keyof BaseTLUiComponents]: BaseTLUiComponents[K] | null
+}>
+
+const TldrawUiComponentsContext = createContext({} as TLUiComponents)
+
+type ComponentsContextProviderProps = {
+	overrides?: TLUiComponents
+	children: any
+}
+
+/** @public */
+export function TldrawUiComponentsProvider({
+	overrides,
+	children,
+}: ComponentsContextProviderProps) {
+	return (
```

Comment:
Can we stick a `overrides = useShallowObjectIdentity(overrides)` in here? That way if people want to write `<Tldraw overrides={{ Thing: MyThing }} />` we won't have re-render loops etc.

Reply:
I think that happens higher up? but yes, can do

Reply:
I'll add it to the editor components too